File: sat_access/sat_access.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import argparse
import urllib.request
import bz2
# import netCDF4 as nc
import xarray as xr
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature
import numpy as np
import pandas as pd
from datetime import datetime, timedelta

def download_and_plot(dl_var, dl_dates, bbox, output_dir, plot_var, overwrite):
    """
    Downloads a NetCDF file from a specified URL and plots a variable as a map.

    Parameters:
    - dl_var: Variable to download (e.g., "SPM", "SPIM", "CHLA").
    - dl_date: Date in YYYY-MM-DD format.
    - bbox: Bounding box as [lon_min, lon_max, lat_min, lat_max].
    - output_dir: Directory to save the downloaded file and plot.
    - overwrite: Whether to overwrite existing files.
    """

    # Download code
    print("Downloading file...")

    # Check dates
    if len(dl_dates) > 2:
        raise ValueError("Please provide only one or two dates for the date range.")

    # Set start and end dates for download
    start_date = datetime.strptime(dl_dates[0], '%Y-%m-%d')
    if len(dl_dates) == 2:
        end_date = datetime.strptime(dl_dates[1], '%Y-%m-%d') 
    else:
        end_date = start_date

    # Iterate over each date in the range
    current_date = start_date
    while current_date <= end_date:
        dl_date = current_date.strftime('%Y-%m-%d')
        dl_date_flat = dl_date.replace("-", "")

        # Get year and day of year for URL
        url_year_doy = f"{dl_date[:4]}/{current_date.strftime('%j')}"

        # Get general URL based on desired variables
        if dl_var.upper() in ["SPM", "SPIM", "CHLA"]:
            url_base = "ftp://ftp.ifremer.fr/ifremer/cersat/products/gridded/ocean-color/atlantic"
        else:
            raise ValueError("Variable not yet available")

        # Get product specifics
        if dl_var.upper() in ["SPM", "SPIM"]:
            file_name = f"{dl_date_flat}-EUR-L4-SPIM-ATL-v01-fv01-OI.nc.bz2"
            url_product = "EUR-L4-SPIM-ATL-v01"
            nc_var_name = "analysed_spim"
            var_label = "SPM [g m-3]"
            nc_file = os.path.join(output_dir, f"{dl_date_flat}-EUR-L4-SPIM-ATL-v01-fv01-OI.nc")
        elif dl_var.upper() == "CHLA":
            file_name = f"{dl_date_flat}-EUR-L4-CHL-ATL-v01-fv01-OI.nc.bz2"
            url_product = "EUR-L4-CHL-ATL-v01"
            nc_var_name = "analysed_chl_a"
            var_label = "chl a [mg m-3]"
            nc_file = os.path.join(output_dir, f"{dl_date_flat}-EUR-L4-CHL-ATL-v01-fv01-OI.nc")
        else:
            raise ValueError("Variable not yet available")

        # Assemble final URL
        url_final = f"{url_base}/{url_product}/{url_year_doy}/{file_name}"
        file_name_full = os.path.join(output_dir, file_name)

        # Fetch file
        if os.path.exists(file_name_full) and not overwrite:
            print(f"{file_name_full} already exists. Set --overwrite True to force the download.")
        else:
            try:
                urllib.request.urlretrieve(url_final, file_name_full)
                print(f"File downloaded at: {file_name_full}")
                # Extract the .bz2 file
                with open(file_name_full, 'rb') as source, open(file_name_full[:-4], 'wb') as dest:
                    dest.write(bz2.decompress(source.read()))
                print(f"File extracted at: {file_name_full[:-4]}")
            except Exception as e:
                print(f"Failed to download or extract {url_final}: {e}")

        # Move to the next day
        current_date += timedelta(days=1)
    
    # plot_var = False 
    print(f"plot_var: {plot_var}, type: {type(plot_var)}")
    # Plotting code
    if plot_var:

        print("Plotting...")

        if len(dl_dates) == 2:
            print("Two dates provided; the last date will be used for plotting.")

        # Open the NetCDF file using xarray
        ds = xr.open_dataset(nc_file)

        # Get date value as a label for plotting
        date_value = ds.time.values[0]  # Extract the numpy.datetime64 value
        date_time_obj = pd.to_datetime(date_value)
        date_label = date_time_obj.strftime('%Y-%m-%d')

        # Subset NetCDF file to desired bounding box
        var_subset = ds[nc_var_name].where(
            (ds.lon >= bbox[0]) & (ds.lon <= bbox[1]) &
            (ds.lat >= bbox[2]) & (ds.lat <= bbox[3]),
            drop=True
        )

        # Create a figure and axes with a Plate Carree projection
        fig = plt.figure(figsize=(10, 6))
        ax = plt.axes(projection=ccrs.PlateCarree())

        # Plot the data
        raster = var_subset.plot(
            ax=ax,
            transform=ccrs.PlateCarree(),
            cmap='viridis',
            add_colorbar=True,
            cbar_kwargs={'label': var_label}
        )

        # Add geospatial features
        ax.add_feature(cfeature.LAND)
        ax.add_feature(cfeature.OCEAN)
        ax.add_feature(cfeature.COASTLINE)
        ax.add_feature(cfeature.BORDERS, linestyle=':')
        ax.gridlines(draw_labels=True)

        # Add a title
        plt.title(f'Map of {nc_var_name} on {date_label}')

        # Show the plot
        plt.show()

    else:
        print("Plotting skipped.")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Download SPM or chl a NetCDF files and plot them as a map.")
    parser.add_argument("--variable", type=str, required=True, help="Variable to download (e.g., 'SPM', 'CHLA')")
    parser.add_argument("--daterange", type=str, nargs='+', required=True, 
                        help="Date range for desired data in YYYY-MM-DD format. Provide only one or two values. Note that if two dates are provided, only the first will be used for plotting.")
    parser.add_argument("--boundingbox", type=float, nargs=4, required=True, help="Bounding box as lon_min lon_max lat_min lat_max")
    parser.add_argument("--outputdir", type=str, required=True, help="Directory to save the downloaded file and plot")
    parser.add_argument("--plot", type=bool, default=False, help="Whether or not to plot the downloaded data. Default = True")
    parser.add_argument("--overwrite", type=bool, default=False, help="Overwrite existing files. Default = False")

    args = parser.parse_args()
    download_and_plot(args.variable, args.daterange, args.boundingbox, args.outputdir, args.plot, args.overwrite)

File: sat_access/sat_access.R
#!/usr/bin/env Rscript


# Libraries ---------------------------------------------------------------

# Check for missing libraries and install them if necessary
if (!all(c("argparse", "ncdf4", "curl", "reshape2", "ggplot2") %in% installed.packages())) {
  install.packages(c("argparse", "ncdf4", "curl", "reshape2", "ggplot2"), repos = "https://cloud.r-project.org/")
}

# Uncomment this line if R cannot find python on Windows:
# options(python_cmd = "C:/Path/To/Your/Python/python.exe")

# Activate libraries
library(argparse) # For parsing arguments from the command line
library(ncdf4)    # For reading NetCDF files
suppressPackageStartupMessages(library(curl)) # For FTP download
library(reshape2) # For data reshaping
library(ggplot2)  # For visualization


# Parse arguments ---------------------------------------------------------

# NB: If you have opened this script in R/RStudio and intend to run it manually, this section can be ignored.

# Create the parser
parser <- ArgumentParser(description = "Download a NetCDF file from FTP and plot a variable as a map.")

# Add arguments
parser$add_argument("-v", "--variable", type = "character", required = TRUE, help = "Surface variable to fetch and plot")
parser$add_argument("-d", "--daterange", type = "character", nargs = '+', required = TRUE, 
                    help = "Date range of the desired variable in YYYY-MM-DD. Provide only one or two values. 
                    Note that if two dates are provided, only the first will be used for plotting.")
parser$add_argument("-od", "--outputdir", type = "character", required = TRUE, help = "Location to save the NetCDF file and output image")
parser$add_argument("-ov", "--overwrite", type = "logical", default = FALSE, 
                    help = "Whether to overwrite an existing file or not. Default = FALSE")
parser$add_argument("-p", "--plot", type = "logical", default = FALSE, 
                    help = "Whether or not to plot the downloaded data. Default = False")
parser$add_argument("-bbox", "--boundingbox", nargs = 4, type = "double", required = FALSE,
                    help = "The bounding box for plotting. Must be given as: lonmin, lonmax, latmin, latmax")

# Create the function
args <- parser$parse_args()


# The function to call ----------------------------------------------------

# Main function to download and plot
download_and_plot <- function(dl_var, dl_dates, bbox, output_dir, plot_var, overwrite) {
  
  ## Download code -----------------------------------------------------------

  # Check date range
  if(length(dl_dates) > 2){
    stop("Please provide only one or two dates for the date range.")
  }
  
  message("Downloading files...")
  
  # Set start and end dates for download
  start_date <- as.Date(dl_dates[1])
  if(length(dl_dates) == 2){
    end_date <- as.Date(dl_dates[2])
  } else{
    end_date <- start_date
  }
  
  # Iterate over each date in the range
  current_date <- start_date
  while(current_date <= end_date){
    
    # Prep date strings
    dl_date <- current_date
    dl_date_flat <- gsub("-", "", dl_date)
    url_year_doy <- paste0(substr(dl_date, start = 1, stop = 4),"/",strftime(as.Date(dl_date), format = "%j"))
    
    # Get general URL based on desired variables
    if(toupper(dl_var) %in% c("SPM", "SPIM", "CHLA")){
      url_base <- "ftp://ftp.ifremer.fr/ifremer/cersat/products/gridded/ocean-color/atlantic"
    } else {
      stop("Variable not yet available")
    }
    
    # Get product specifics
    if(toupper(dl_var) %in% c("SPM", "SPIM")){
      file_name <- paste0(dl_date_flat,"-EUR-L4-SPIM-ATL-v01-fv01-OI.nc.bz2")
      url_product <- "EUR-L4-SPIM-ATL-v01"
      nc_file <- file.path(output_dir, paste0(dl_date_flat,"-EUR-L4-SPIM-ATL-v01-fv01-OI.nc"))
      nc_var_name <- "analysed_spim"
      var_label <- "SPM [g m-3]"
    } else if(toupper(dl_var) == "CHLA"){
      file_name <- paste0(dl_date_flat,"-EUR-L4-CHL-ATL-v01-fv01-OI.nc.bz2")
      url_product <- "EUR-L4-CHL-ATL-v01"
      nc_file <- file.path(output_dir, paste0(dl_date_flat,"-EUR-L4-CHL-ATL-v01-fv01-OI.nc"))
      nc_var_name <- "analysed_chl_a"
      var_label <- "chl a [mg m-3]"
    } else {
      stop("Variable not yet available")
    }
    
    # Assemble final URL
    url_final <- paste(url_base, url_product, url_year_doy, file_name, sep = "/")
    file_name_full <- file.path(output_dir, file_name)
    
    # Fetch file
    if(file.exists(file_name_full) & !overwrite){
      
      message(paste0(file_name_full," already exists. Set --overwrite TRUE to force the download."))
      
    } else {
      
      # Ensure output directory exists
      if(!dir.exists(output_dir)){
        dir.create(output_dir, recursive = TRUE)
      }
      
      # Download
      curl::curl_download(url_final, destfile = file_name_full)
      message(paste0("File downloaded at: ",file_name_full))
      
      # Unzip
      system(paste("bunzip2 -k -f", file_name_full))
      message("File unzipped at: ", gsub(".bz2","",file_name_full))
    }
    
    # Move to the next day
    current_date <- current_date + 1
  }


  ## Plotting code -----------------------------------------------------------

  if(!plot_var){
    message("Plotting skipped as per user request.")
  } else {
    message("Plotting...")
    
    # User warning about date range
    if(length(dl_dates) == 2){
      message("Two dates provided; the last date will be used for plotting.")
    }
    
    # Set plot name
    plot_name <- file.path(output_dir, paste0(nc_var_name,"_",dl_date,".png"))
    
    # Open the NetCDF file
    nc_data <- nc_open(nc_file)
    
    # Extract longitude, latitude, and the specified variable
    lon <- ncvar_get(nc_data, "lon")
    lat <- ncvar_get(nc_data, "lat")
    time <- ncvar_get(nc_data, "time")
    var <- ncvar_get(nc_data, nc_var_name)
    
    # Close the NetCDF file
    nc_close(nc_data)
    
    # Reshape data for ggplot
    var_df <- melt(var)
    names(var_df) <- c("lon_idx", "lat_idx", "value")
    var_df$lon <- lon[var_df$lon_idx]
    var_df$lat <- lat[var_df$lat_idx]
    
    if(is.null(bbox)){
      bbox <- c(
        min(lon, na.rm = TRUE),
        max(lon, na.rm = TRUE),
        min(lat, na.rm = TRUE),
        max(lat, na.rm = TRUE)
      )
      message(paste("No bounding box provided, using full extent:", paste(round(bbox, 4), collapse = ", ")))
    }
    
    # Filter df to bounding box
    var_df_sub <- var_df[var_df$lon >= bbox[1] & var_df$lon <= bbox[2],]
    var_df_sub <- var_df_sub[var_df_sub$lat > bbox[3] & var_df_sub$lat <= bbox[4], ]
    
    # Get date for plot label
    plot_date <- as.Date(as.POSIXct(time, origin = "1998-01-01"))
    
    # Plot using ggplot2
    p <- ggplot(var_df_sub, aes(x = lon, y = lat, fill = value)) +
      geom_tile() +
      scale_fill_viridis_c(var_label) +
      labs(title = paste("Map of", nc_var_name, "on", plot_date),
           x = "Longitude", y = "Latitude") +
      theme_minimal() +
      theme(legend.position = "bottom")
    ggsave(filename = plot_name, plot = p)
    message(paste0("Image saved at: ",plot_name))
  }
}


# Parse the args ----------------------------------------------------------

# NB: If you have opened this script in R/RStudio and intend to run it manually, 
# uncomment the following chunk of code and add your desired values directly:
# args <- list(
#   variable = "SPM",
#   daterange = c("2023-01-01", "2023-01-05"),
#   outputdir = "path/to/output/dir",
#   overwrite = TRUE,
#   plot = TRUE,
#   boundingbox = c(-10, 10, 35, 45)
# )

# Call the main function with parsed arguments
download_and_plot(
  dl_var = args$variable,
  dl_dates = args$daterange,
  output_dir = args$outputdir,
  overwrite = args$overwrite,
  plot_var = args$plot,
  bbox = args$boundingbox
)

